### PR TITLE
Add 2.4-2.6 field types, range search, resource groups, replica ops and row-based insert/upsert

### PR DESCRIPTION
**Pull request by N-E-W-T-O-N**
Brings the SDK up the 2.4–2.6 feature track: four new field types, range search, resource groups and replica operations, and row-based insert/upsert. 23 commits, each self-contained with its own tests.

Every change is verified against the **full CI matrix** — `v2.3.22`, `v2.4.23`, `v2.5.20`, `v2.6.4` — at **235 passed / 1 skipped / 0 failed** per version. The single skip is the pre-existing `Query_with_time_travel`.

## Data types

| Type | Notes |
|---|---|
| `Int8Vector` | Full insert / read / search path. Only `HNSW` and `AUTOINDEX` accept it — `FLAT` is rejected server-side with `data type Int8Vector can't build with this index`, established by probing rather than assumed. |
| `BFloat16Vector` | Needs a dedicated `BFloat16` struct: search dispatches on element type, and `ReadOnlyMemory<float>` already means `FloatVector`. There is no BCL bfloat16, and `BitConverter.SingleToUInt32Bits` is unavailable on netstandard2.0/net462, so the conversion uses round-to-nearest-even over an unsafe reinterpret. |
| `Geometry` | Plus the `RTREE` index. |
| `Timestamptz` | Travels as ISO 8601 text. |

## Search and indexing

- **Strongly typed range search** — `SearchParameters.Radius` / `RangeFilter` instead of hand-built params JSON (#133). Typed values win over any matching dictionary entry.
- **`SPARSE_WAND`** index type.

## Resource groups and replicas

`CreateResourceGroup`, `ListResourceGroups`, `DescribeResourceGroup`, `UpdateResourceGroups`, `DropResourceGroup`, `TransferReplica`, and now `GetReplicas` and `LoadBalance`.

Dropping a group needs its limits zeroed first, otherwise Milvus answers `resource group's limits node num is not 0`; that is documented on `DropResourceGroupAsync` and handled in test teardown.

Two server behaviours are pinned down by tests rather than left as folklore, both verified on all four versions:

- **`GetReplicas.with_shard_nodes` is inverted.** The proto says `ShardReplica.node_ids` is *"set only for GetReplicas() if with_shard_nodes is true"*, but `true` returns **empty** per-shard lists and `false` returns **populated** ones. `GetReplicasAsync` therefore defaults the parameter to `false` and passes it through **unnegated**, so a server-side fix keeps working and only the doc note goes stale. A canary test fails the day that happens. The replica-level node list is populated either way.
- **`GetReplicas` on an unloaded collection differs by release line.** From 2.4 it returns an empty list; 2.3 throws `replica not found`. The test asserts both branches instead of skipping on 2.3.

`LoadBalanceAsync` takes nullable target-node and segment lists, since leaving either out asks Milvus to choose — the common case. On a single-query-node deployment it succeeds and does nothing, so the happy-path test is explicit that it verifies only that the request is well formed.

## Row-based insert and upsert

Milvus is column-oriented on the wire, so these pivot `IDictionary<string, object?>` rows into columns client-side at the cost of one `DescribeCollection`. Integer widening is handled (a hand-built dictionary rarely bothers with the `L` suffix), dynamic fields are inferred from the first non-null value, and errors name the offending field.

`UpsertAsync` differs from insert in one way: it must carry the primary key even when the field is `autoId`, because the server needs it to identify the row being replaced — hence the `includeAutoId` flag on the shared pivot.

## Bug fix

`EnableDynamicField` was never read back in `DescribeCollection`, so a round-tripped schema silently lost it.

## Proto

Pinned to **`go-api/v2.6.23`**. The submodule had drifted onto the master/3.0 track, which carries types (`Decimal`, `UUID`, `Date`, `Time`) that no GA server implements — `Decimal` additionally has an unmerged draft PR proposing a different wire format. Every symbol this SDK needs was confirmed present at the pin.

## Test infrastructure

`xunit.runner.visualstudio` to 4.0.0. `GitHubActionsTestLogger` stays at 2.4.1 with an inline comment — 3.0.5 breaks the build with CS1705.

`FlushAll` is cluster-wide, so `DataTests` is marked `DisableParallelization` — it was racing with collection drops in other classes.

### CODE REVIEW

File: Milvus.Client/MilvusCollection.Rows.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    private static FieldData BuildColumn(FieldSchema field, object?[] values)
+        => field.DataType switch
+        {
+            MilvusDataType.Bool => FieldData.Create(field.Name, Cast<bool>(field.Name, values)),
+            MilvusDataType.Int8 => FieldData.Create(field.Name, values.Select(v => (sbyte)ToInt64(v)).ToList()),
+            MilvusDataType.Int16 => FieldData.Create(field.Name, values.Select(v => (short)ToInt64(v)).ToList()),
+            MilvusDataType.Int32 => FieldData.Create(field.Name, values.Select(v => (int)ToInt64(v)).ToList()),
+            MilvusDataType.Int64 => FieldData.Create(field.Name, values.Select(ToInt64).ToList()),
```
Comment (yhmo) [inline Milvus.Client/MilvusCollection.Rows.cs]: Milvus.Client/MilvusCollection.Rows.cs line:215
Medium ---- Row-based insert/upsert does not actually honor nulls for nullable scalar fields, despite the XML doc promising "a field may be omitted or null when the schema marks it nullable". If a nullable Int64/Int32/etc. field is present in some rows but null (or omitted) in others, values[i] stays null and ToInt64(null) throws "Null is not valid for a non-nullable integer field." — a message that is doubly wrong for a nullable field. Only the VarChar path (ToNullableString) handles nulls; nullable numeric/bool/float/double fields will throw instead of being sent as null via ValidData. Consider emitting nullable FieldData (e.g. FieldData.Create<long?> with ValidData) when field.Nullable, and add a test inserting a null into a nullable numeric column.

File: Milvus.Client/MilvusCollection.Rows.cs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                $"Dynamic field '{name}' has unsupported value type '{sample.GetType()}'. Supported types are " +
+                "bool, integers, floating point numbers and strings.")
+        };
+    }
+
+    private static FieldData BuildColumn(FieldSchema field, object?[] values)
+        => field.DataType switch
+        {
+            MilvusDataType.Bool => FieldData.Create(field.Name, Cast<bool>(field.Name, values)),
+            MilvusDataType.Int8 => FieldData.Create(field.Name, values.Select(v => (sbyte)ToInt64(v)).ToList()),
```
Comment (yhmo) [inline Milvus.Client/MilvusCollection.Rows.cs]: Milvus.Client/MilvusCollection.Rows.cs line:212
Low ---- The narrowing integer casts here are unchecked: (sbyte)ToInt64(v), (short)ToInt64(v) and (int)ToInt64(v) silently wrap values that fall outside the target type range (e.g. 300 for an Int8 field becomes 44) instead of reporting an error. The widening direction is deliberately supported and documented, but the narrowing direction silently corrupts data on out-of-range input. Consider range-checking (or using checked casts) and naming the offending field and row in the exception.

File: Milvus.Client/MilvusCollection.Rows.cs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    /// </summary>
+    private static FieldData BuildDynamicColumn(string name, object?[] values)
+    {
+        object? sample = values.FirstOrDefault(v => v is not null);
+
+        return sample switch
+        {
+            null => FieldData.CreateVarChar(name, values.Select(_ => (string?)null).ToList(), isDynamic: true),
+            bool => FieldData.Create(name, Cast<bool>(name, values), isDynamic: true),
+            sbyte or short or int or long => FieldData.Create(name, values.Select(ToInt64).ToList(), isDynamic: true),
```
Comment (yhmo) [inline Milvus.Client/MilvusCollection.Rows.cs]: Milvus.Client/MilvusCollection.Rows.cs line:199
Low ---- Inconsistency between the dynamic-column inference pattern and ToInt64: the pattern matches only sbyte/short/int/long, but ToInt64 below explicitly also accepts byte/ushort/uint. A dynamic field whose first non-null value is a byte (or ushort/uint) therefore falls through to the "unsupported value type" branch even though ToInt64 can convert it. Consider widening the pattern to byte/ushort/uint (or reusing the same set) so inference and conversion stay in sync.

File: Milvus.Client/Int8VectorFieldData.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        };
+
+        ArrayPool<byte>.Shared.Return(bytes);
+
+        return result;
+    }
+
+    internal override object GetValueAsObject(int index)
+        => throw new NotSupportedException("Dynamic vector fields are not supported");
+}
```
Comment (yhmo) [inline Milvus.Client/Int8VectorFieldData.cs]: Milvus.Client/Int8VectorFieldData.cs line:83
Low ---- This new file is missing a trailing newline at EOF. The repo .editorconfig sets insert_final_newline = true for [*.cs], so the file deviates from the enforced formatting (sibling BFloat16VectorFieldData.cs and Float16VectorFieldData.cs both end with a newline).

File: Milvus.Client/MilvusDataType.cs [hunk 5]
```diff
@@ -83,8 +93,34 @@ public enum MilvusDataType
...
+
+    /// <summary>
+    /// Int8Vector.
+    /// </summary>
+    Int8Vector = 105,
+
+    /// <summary>
+    /// Array of Vector.
+    /// </summary>
+    ArrayOfVector = 106,
```
Comment (yhmo) [inline Milvus.Client/MilvusDataType.cs]: Milvus.Client/MilvusDataType.cs line:114
Low ---- ArrayOfVector/ArrayOfStruct/Struct are added to the public enum but have no SDK support: EnsureDataType doesn't map them, there is no FieldSchema.Create convenience or FieldData path for them, and ToGrpcFieldData would throw "DataType Error:..., not supported" at insert time. The proto marks them "internal types, not used in user interface". Exposing them publicly can mislead callers into thinking they are usable. Consider dropping them from the public enum (or documenting them as unsupported) until the SDK actually implements them.

File: Milvus.Client/MilvusCollection.Rows.cs (hunk 3 above)
Comment (yhmo) [inline Milvus.Client/MilvusCollection.Rows.cs]: Milvus.Client/MilvusCollection.Rows.cs line:199
Medium ---- Dynamic fields are per-row optional in Milvus (each row's JSON may carry a different set of extra keys), but this pivot path throws for numeric/bool dynamic fields that appear in only some rows. In AddDynamicColumns (line 180), a row that lacks the key leaves values[i] = null, and then ToInt64(null) / ToDouble(null) / Cast<bool> throw "Null is not valid for a non-nullable integer field." — the string case (CreateVarChar) tolerates nulls, so behavior is inconsistent. A batch where a numeric dynamic field is present in row 1 but absent in row 2 fails instead of omitting the key for row 2. Consider treating null as "key absent" for dynamic columns (or documenting that numeric dynamic fields must be present in every row), and add a test with a numeric dynamic field present in only some rows.

File: Milvus.Client.Tests/SearchQueryTests.cs [hunk 6]
```diff
@@ -570,6 +615,399 @@ await float16VectorCollection.WaitForCollectionLoadAsync(
     }
 #endif
 
+    // Milvus 2.6 only accepts HNSW (and AutoIndex) for Int8Vector fields; FLAT, IVF_*, DISKANN and
```
Comment (yhmo) [inline Milvus.Client.Tests/SearchQueryTests.cs]: Milvus.Client.Tests/SearchQueryTests.cs line:618
Low ---- This comment ("Milvus 2.6 only accepts HNSW (and AutoIndex) for Int8Vector fields...") describes the Int8Vector index restriction, but it sits directly above BFloat16_conversion_round_trips instead of above Search_int8_vector, which is the test that actually uses HNSW/AutoIndex with Int8Vector. A reader of the BFloat16 test could be misled into thinking the restriction applies to bfloat16. Consider moving it above the Search_int8_vector [Theory].

File: Milvus.Client/MilvusCollection.Rows.cs (hunk 1 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/MilvusCollection.Rows.cs]: Fixed in 44a49ef. You were right that the machinery already existed — building the column as `List<T?>` is what makes `FieldData` emit the `valid_data` mask, so the fix was to consult `field.Nullable` rather than add anything new.

Columns are now built nullable when the schema says so, and a null in a column that is *not* nullable is reported against the specific field and row instead of producing that misleading message. Tests cover all three shapes for a nullable numeric column: an explicit `null`, an omitted key, and a real value, gated at 2.5 where nullable fields were introduced.

File: Milvus.Client/MilvusCollection.Rows.cs (hunk 2 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/MilvusCollection.Rows.cs]: Fixed in 44a49ef. `Int8`, `Int16` and `Int32` are now range-checked and the exception names the field, the row and the permitted range.

Worth noting `Float` had the identical bug wearing a different hat: a finite `double` outside float's range narrows to infinity rather than failing, so that is checked too. Tests cover 300 into an Int8 field plus the `sbyte` boundaries (127 / -128) still going through.

File: Milvus.Client/MilvusCollection.Rows.cs (hunk 3 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/MilvusCollection.Rows.cs]: Fixed in 44a49ef. The inference pattern and the converters now share one accepted set including `byte`/`ushort`/`uint`, and there is a test inserting dynamic fields whose first values are `(byte)7`, `(ushort)8` and `9u`.

File: Milvus.Client/Int8VectorFieldData.cs (hunk 4 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/Int8VectorFieldData.cs]: Fixed in 8c9e23b.

File: Milvus.Client/MilvusDataType.cs (hunk 5 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/MilvusDataType.cs]: Fixed in 691dc55 — dropped rather than documented. Agreed on the reasoning: a public enum member that no code path accepts only misleads callers, and the proto marks them internal. They can come back with the support that makes them mean something.

File: Milvus.Client/MilvusCollection.Rows.cs (hunk 3 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client/MilvusCollection.Rows.cs]: Fixed in 44a49ef, and thank you for this one — it could not be fixed in the pivot alone, which I would not have found without the pointer.

Making dynamic columns nullable immediately failed with `InvalidCastException: Unable to cast List<Nullable<Int64>> to IReadOnlyList<Int64>` at `FieldData.GetValueAsObject`. Dynamic fields never reach `ToGrpcFieldData` — `PopulateData` pulls them value-by-value through `GetValueAsObject` and packs them into the `$meta` JSON column, and that method cast `Data` to a list of the non-nullable element type. So two changes were needed:

- `GetValueAsObject` now boxes `Data[index]` instead of casting the list, since a `List<T?>` is not an `IReadOnlyList<T>`. Return type widened to `object?`.
- `PopulateData` omits a null-valued dynamic key rather than writing `"key": null` — an explicit JSON null makes the key *exist* with a null value, which is not the same as the row not having it.

Side effect: this also fixes the column-based API, where `FieldData.Create<long?>(..., isDynamic: true)` previously threw the same `InvalidCastException`. Test inserts a numeric dynamic field present in one of two rows.

File: Milvus.Client.Tests/SearchQueryTests.cs (hunk 6 above)
Comment (N-E-W-T-O-N) [reply Milvus.Client.Tests/SearchQueryTests.cs]: Fixed in 8c9e23b — moved above the `Search_int8_vector` theory, and extended slightly to say that the restriction is why the theory only covers those two index types.
